### PR TITLE
Early stopping

### PR DESCRIPTION
**Pull request by owain-Biddulph**

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -73,12 +74,20 @@ def generate_subspectrogram(duration_s=config.SUBSPECTROGRAM_DURATION_S, fft_rat
...
+                model, val_spectrograms, val_labels, val_loss, val_accuracy, epoch=epoch+1))
 
     utils.save_model(model, epoch)
-    utils.save_and_display_loss_through_epochs(epoch_range, loss_history, model.name)
+    utils.save_and_display_loss_through_epochs(epoch_range, loss_history, model.name, "training loss")
+
+    # Plot val accuracies
+    val_accuracy_range = [i for i in epoch_range if i % 10 == 0 or i == config.NB_EPOCHS - 1]
+    utils.save_and_display_loss_through_epochs(val_accuracy_range, val_accuracies, model.name, "validation accuracy")
+    best_epoch = (len(val_accuracies) - list(reversed(val_accuracies)).index(min(val_accuracies)) - 1) * 10
```
Comment (Antoine-marchais) [inline train.py]: Considering how jumpy the accuracy can be, I think we are better off checking the curve to be sure of the minima. We might want to plot accuracy as well

Comment (owain-Biddulph) [reply train.py]: Sorry that is actually the loss, forgot to change the variable name. 
The best epoch is just an indication, validation loss is plotted (l. 88). We could do both, I’ll add that.

File: train.py [hunk 2]
```diff
@@ -73,12 +75,23 @@ def generate_subspectrogram(duration_s=config.SUBSPECTROGRAM_DURATION_S, fft_rat
...
         checkpoint_manager.save()
 
-    print("======================== evaluation on validation data =========================")
-    # test model on validation set
-    utils.test_model(model, val_spectrograms, val_labels, val_loss, val_accuracy)
+        if epoch % 10 == 0 or epoch == config.NB_EPOCHS - 1:
+            print("======================== evaluation on validation data =========================")
+            # test model on validation set
+            this_val_accuracy, this_val_loss = utils.test_model(
+                model, val_spectrograms, val_labels, val_loss, val_accuracy, epoch=epoch+1)
```
Comment (quentinverlhac) [inline train.py]: Why `epoch+1` ? Everywhere else we use `epoch`, starting at 0.

File: utils.py [hunk 3]
```diff
@@ -173,13 +175,13 @@ def average_predictions(full_spectrogram, model):
     return tf.add_n(all_predictions)/len(segmented_spectro)
 
 
-def save_and_display_loss_through_epochs(epoch_range, loss, model_name):
+def save_and_display_loss_through_epochs(epoch_range, loss, model_name, y_label):
```
Comment (quentinverlhac) [inline utils.py]: You should change the name of the function `save_and_display_metric_through_epochs`

File: train.py [hunk 4]
```diff
@@ -73,12 +75,23 @@ def generate_subspectrogram(duration_s=config.SUBSPECTROGRAM_DURATION_S, fft_rat
         checkpoint.epoch.assign_add(1)
         checkpoint_manager.save()
 
-    print("======================== evaluation on validation data =========================")
-    # test model on validation set
-    utils.test_model(model, val_spectrograms, val_labels, val_loss, val_accuracy)
+        if epoch % 10 == 0 or epoch == config.NB_EPOCHS - 1:
```
Comment (quentinverlhac) [inline train.py]: The epoch period (10 here) could be a config variable

File: train.py [hunk 5]
```diff
@@ -73,12 +75,23 @@ def generate_subspectrogram(duration_s=config.SUBSPECTROGRAM_DURATION_S, fft_rat
...
 
     utils.save_model(model, epoch)
-    utils.save_and_display_loss_through_epochs(epoch_range, loss_history, model.name)
+    utils.save_and_display_loss_through_epochs(epoch_range, loss_history, model.name, "training loss")
+
+    # Plot val accuracies
+    val_range = [i for i in epoch_range if i % 10 == 0 or i == config.NB_EPOCHS - 1]
+    utils.save_and_display_loss_through_epochs(val_range, val_accuracies, model.name, "validation accuracy")
+    utils.save_and_display_loss_through_epochs(val_range, val_losses, model.name, "validation loss")
+    best_epoch = (len(val_accuracies) - list(reversed(val_accuracies)).index(min(val_accuracies)) - 1) * 10
```
Comment (quentinverlhac) [inline train.py]: Why not just `val_loss.index(min(val_loss)) * 10` ?

Comment (owain-Biddulph) [reply train.py]: Because I want the last one, not the first.

File: utils.py (hunk 3 above)
Comment (owain-Biddulph) [reply utils.py]: Yes

File: train.py (hunk 4 above)
Comment (owain-Biddulph) [reply train.py]: True

File: train.py (hunk 2 above)
Comment (owain-Biddulph) [reply train.py]: yeah, just seems weird
